### ISSUE TITLE
[docs] critical API documentation is missing

### ISSUE THREAD

**Issue by stas00** [microsoft/DeepSpeed#776]
It looks like things kind of work, except not quite right, and there are a lot of subtle nuances that are so hard to miss when integrating DeepSpeed. I think all these should be made loud and clear - and perhaps a simple full example of a training loop would help, including showing commented out code where the original training code is removed to do it the DeepSpeed-way.

As I am trying to figure out how to make `gradient_accumulation_steps` work correctly I'm finding all kinds of things I have missed when integrating DeepSpeed into HF Trainer. I will post them here as I find such things:

1. engine's `backward` returns `loss`, which it modifies under `gradient_accumulation_steps > 1` but this is undocumented. 
   * neither in the docstring API https://github.com/microsoft/DeepSpeed/blob/e60e92eb0a06673748c4cb63fbcf713ddd12fc22/deepspeed/runtime/engine.py#L852-L858
   * nor main docs: https://www.deepspeed.ai/getting-started/#training
   
2. Also it's not documented that the "client" shouldn't scale loss by `gradient_accumulation_steps` since Deepspeed does it in `backward`.
   
3. the fact that `lr_scheduler.step` happens inside engine's `step` is not documented in the API
   * https://github.com/microsoft/DeepSpeed/blob/e60e92eb0a06673748c4cb63fbcf713ddd12fc22/deepspeed/runtime/engine.py#L993-L996
   * but is documented in https://www.deepspeed.ai/getting-started/#training
   * it might be a good idea to also add an explicit - make sure to remove `lr_scheduler.step()` from your code if using DeepSpeed's scheduler.

**Comment by janEbert** [comment]
Adding to this:
5. Remove `model = model.cuda()` and `model = model.half()` calls.
6. Do keep `data = data.cuda()` and optionally add `data = data.half()` calls.
7. Substitute `torch.nn.utils.clip_grad_norm_(params, X)` with `"gradient_clipping": X` in the DeepSpeed config.
8. Guard parallel I/O using parallel primitives (except DeepSpeed checkpointing which is already nicely documented [in the code](https://github.com/microsoft/DeepSpeed/blob/14a50c68c33968f9cd7fefd29e8fe72d4be200e8/deepspeed/runtime/engine.py#L1625-L1628) and [at deepspeed.ai](https://github.com/microsoft/DeepSpeed/blame/14a50c68c33968f9cd7fefd29e8fe72d4be200e8/docs/_tutorials/getting-started.md#L134-L136)).
9. Use `torch.distributed` methods for parallel primitives (e.g. `torch.distributed.get_world_size()`, `torch.distributed.get_rank()`, `int(os.environ['LOCAL_RANK'])`, ...).

And a bit less interesting (though one of these should probably be adjusted instead of documented):
10. In `deepspeed.initialize`, the `optimizer` kwarg has precedence over the `"optimizer"` key in the DeepSpeed config.
11. In `deepspeed.initialize`, the `"scheduler"` key in the DeepSpeed config has precedence over the `lr_scheduler` kwarg.

**Comment by stas00** [comment]
Thank you for the additions, @janEbert!

> 6. Do keep `data = data.cuda()` and optionally add `data = data.half()` calls.

And since fp32 is being worked on https://github.com/microsoft/DeepSpeed/pull/1004, this could be dynamic.
```
inputs = inputs.to(dtype=next(iter(model.parameters())).dtype)
```
so it can automatically handle the case of `model.half()` (fp16) or not (fp32) down the road.

**Comment by janEbert** [comment]
Oh nice, that input conversion is even better! Happy to hear FP32 ZeRO 3 is getting PRed as well. :)

**Comment by janEbert** [comment]
Actually I just noticed, so long as the device is set correctly (that is, with `torch.cuda.set_device(local_rank)`), `model = model.cuda()` can be left in and `model = model.half()` can be left in either way.

See [device setup](https://github.com/microsoft/DeepSpeed/blob/41ab660b5df3567966935fe8ac3672497bc8690a/deepspeed/runtime/engine.py#L504-L505) and [model conversion](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/engine.py#L583-L588) code.

**Comment by stas00** [comment]
Based on my understanding this is all taken care of internally by DeepSpeed, so you want to skip both of those calls if your training loop is running under Deepspeed - this is what we do in HF Trainer.

`model = model.cuda()` will OOM anyway if your model is larger than gpu0.

**Comment by janEbert** [comment]
Yea I guess it's better to just give your concrete advice; who knows how the internals will change with the speed this is moving at. :)
